packages/synapse-react-client/src/components/TableColumnSchemaEditor/Validators/ColumnModelValidator.ts

```diff
@@ -0,0 +1,145 @@
+            .max(MAX_STRING_SIZE)
+            .optional(),
+        ),
+      maximumListLength: z
+        .union([optionalStringSchema, z.number()])
+        .pipe(
+          z.coerce
+            .number()
+            .finite()
+            .int()
+            .min(1)
+            .max(MAX_LIST_LENGTH)
+            .optional(),
+        ),
+      enumValues: z.array(z.coerce.string()).optional(),
+      jsonSubColumns: z.array(jsonSubColumnModelZodSchema).optional(),
+      facetType: facetTypeSchema.optional(),
+    }),
+  )
+  .refine(data => data.maximumSize == null || canHaveSize(data.columnType), {
+    message: 'Size is not allowed for this column type',
+    path: ['maximumSize'],
+  })
+  .refine(
+    data =>
+      data.maximumListLength == null || canHaveMaxListLength(data.columnType),
+    {
+      message: 'Maximum list length is not allowed for this column type',
+      path: ['maximumListLength'],
+    },
+  )
+  .refine(
+    data => {
+      if (data.jsonSubColumns != null) {
+        return data.columnType === ColumnTypeEnum.JSON
+      }
+      return true
+    },
+    {
+      message: 'Only JSON columns can have JSON sub-columns.',
+      path: ['columnType'],
+    },
+  )
+  .transform((data, ctx) => {
+    if (data.defaultValue != null) {
+      // Validate and transform the defaultValue based on the columnType
+      const defaultValueSchema = getDefaultValueValidator(data.columnType)
+      const result = defaultValueSchema.safeParse(data.defaultValue)
+      let transformedValue: string | undefined
+      if (result.success) {
+        transformedValue = result.data
+      } else {
+        result.error.issues.forEach(issue => {
+          ctx.addIssue({ ...issue, path: ['defaultValue'] })
+        })
+        transformedValue = undefined
+      }
+      return { ...data, defaultValue: transformedValue }
+    }
+    return omit(data, 'defaultValue')
+  })
+  .transform((data, ctx) => {
+    if (data.enumValues != null) {
+      // Validate and transform the defaultValue based on the columnType
+      const enumValuesSchema = getEnumValuesValidator(data.columnType)
+      const result = enumValuesSchema.safeParse(data.enumValues)
+      let transformedValue: string[] | undefined
+      if (result.success) {
+        transformedValue = result.data
+      } else {
+        result.error.issues.forEach(issue => {
+          ctx.addIssue({ ...issue, path: ['enumValues'] })
+        })
+        transformedValue = undefined
+      }
+      return { ...data, enumValues: transformedValue }
+    }
+    return omit(data, 'enumValues')
+  })
+
+export const columnModelFormDataZodSchema = z.array(columnModelZodSchema)
+
+export function validateColumnModelFormData(
+  formData: SetOptional<
+    ColumnModelFormData,
+    'isOriginallyDefaultColumn' | 'isSelected'
+  >[],
+): SafeParseReturnType<
+  SetOptional<
+    ColumnModelFormData,
+    'isOriginallyDefaultColumn' | 'isSelected'
+  >[],
+  SetOptional<ColumnModel, 'id'>[]
```

Comment:
Small note but pretty cool: this line denotes that the _specified_ return type of the schema's `parse`/`safeParse` functions is our `ColumnModel` object.

The _actual_ return type is inferred based on the Zod schema.

If the specified type and the actual types don't match, then TypeScript returns a compiler error.

So with just this line, we basically have a compile-time test that our schema actually returns a `ColumnModel` object.